ingestion/src/metadata/utils/datalake_utils.py

```diff
@@ -0,0 +1,376 @@
+    try:
+        data = json.loads(json_text)
+    except json.decoder.JSONDecodeError:
+        logger.debug("Failed to read as JSON object trying to read as JSON Lines")
+        data = [json.loads(json_obj) for json_obj in json_text.strip().split("\n")]
+    return dataframe_to_chunks(json_normalize(data, sep=COMPLEX_COLUMN_SEPARATOR))
+
+
+@singledispatch
+def read_csv_dispatch(config_source: Any, **kwargs):
+    raise NotImplementedError(
+        f"Didn't Implement {config_source.__class__.__name__} for CSV"
+    )
+
+
+@singledispatch
+def read_tsv_dispatch(config_source: Any, **kwargs):
+    raise NotImplementedError(
+        f"Didn't Implement {config_source.__class__.__name__} for TSV"
+    )
+
+
+@singledispatch
+def read_avro_dispatch(config_source: Any, **kwargs):
+    raise NotImplementedError(
+        f"Didn't Implement {config_source.__class__.__name__} for AVRO"
+    )
+
+
+@singledispatch
+def read_parquet_dispatch(config_source: Any, **kwargs):
+    raise NotImplementedError(
+        f"Didn't Implement {config_source.__class__.__name__} for PARQUET"
+    )
+
+
+@singledispatch
+def read_json_dispatch(config_source: Any, **kwargs):
+    raise NotImplementedError(
+        f"Didn't Implement {config_source.__class__.__name__} for JSON"
+    )
+
+
+@read_csv_dispatch.register
+def _(_: GCSConfig, key: str, bucket_name: str, **kwargs):
+    """
+    Read the CSV file from the gcs bucket and return a dataframe
+    """
+    path = f"gs://{bucket_name}/{key}"
+    return read_from_pandas(path=path, separator=CSV_SEPARATOR)
+
+
+@read_csv_dispatch.register
+def _(_: S3Config, key: str, bucket_name: str, client, **kwargs):
+    path = client.get_object(Bucket=bucket_name, Key=key)["Body"]
+    return read_from_pandas(path=path, separator=CSV_SEPARATOR)
+
+
+@read_csv_dispatch.register
+def _(config_source: AzureConfig, key: str, bucket_name: str, **kwargs):
+    storage_options = return_azure_storage_options(config_source)
+    path = f"abfs://{bucket_name}@{storage_options.get('account_name')}.dfs.core.windows.net/{key}"
+
+    return read_from_pandas(
+        path=path, separator=CSV_SEPARATOR, storage_options=storage_options
+    )
+
+
+@read_tsv_dispatch.register
+def _(_: GCSConfig, key: str, bucket_name: str, **kwargs):
+    """
+    Read the TSV file from the gcs bucket and return a dataframe
+    """
+    path = f"gs://{bucket_name}/{key}"
+    return read_from_pandas(path=path, separator=TSV_SEPARATOR)
+
+
+@read_tsv_dispatch.register
+def _(_: S3Config, key: str, bucket_name: str, client, **kwargs):
+    path = client.get_object(Bucket=bucket_name, Key=key)["Body"]
+    return read_from_pandas(path=path, separator=TSV_SEPARATOR)
+
+
+@read_tsv_dispatch.register
+def _(config_source: AzureConfig, key: str, bucket_name: str, **kwargs):
+    storage_options = return_azure_storage_options(config_source)
+    path = f"abfs://{bucket_name}@{storage_options.get('account_name')}.dfs.core.windows.net/{key}"
```

Comment:
this path building is repeated in a few places. We might be able to centralise it